core/lib/zksync_core/src/api_server/web3/namespaces/zks.rs

```diff
@@ -620,4 +622,30 @@ impl<G: L1GasPriceProvider> ZksNamespace<G> {
     ) -> Result<Vec<Log>, Web3Error> {
         self.state.translate_get_logs(filter).await
     }
+
+    #[tracing::instrument(skip_all)]
+    pub async fn get_proofs_impl(
+        &self,
+        address: Address,
+        keys: Vec<H256>,
+        l1_batch_number: L1BatchNumber,
+    ) -> Result<Vec<TreeEntryWithProof>, Web3Error> {
+        const METHOD_NAME: &str = "get_proofs";
+
+        let hashed_keys = keys
+            .into_iter()
+            .map(|key| StorageKey::new(AccountTreeId::new(address), key).hashed_key_u256())
+            .collect();
+
+        let keys = self
+            .state
+            .tree_api
+            .as_ref()
+            .ok_or(Web3Error::GetProofsUnavailable)?
+            .get_proofs(l1_batch_number, hashed_keys)
+            .await
+            .map_err(|err| internal_error(METHOD_NAME, err))?;
```

Comment:
A potential issue here is that if the tree API is load-balanced via K8s services (makes sense), a query to a certain upstream server may fail because it doesn't have an L1 batch locally while other servers may have it. Not sure how best to deal with it. A tree API server provides its latest processed L1 batch [via a separate endpoint](https://github.com/matter-labs/zksync-era/blob/664ce33622af220a24360f7f11a52a14141c3fdc/core/lib/zksync_core/src/api_server/tree/mod.rs#L92), so we could theoretically use app-level load-balancing aware of the current server states (encapsulated in a `TreeApiClient` implementation). Obviously, this could be dealt with later.

Reply:
yes, can be dealt with later - but indeed need to think about it, especially if we support BlockNumber/BlockId. Would be silly to get latest block number in the DB and then hit a server with no such l1 batch and return error :)